pkg/controllers/machine/controller.go

```diff
@@ -0,0 +1,201 @@
+		}
+		if err := c.kubeClient.Status().Patch(ctx, statusCopy, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(multierr.Append(errs, err))
+		}
+	}
+	return result.Min(results...), errs
+}
+
+func (c *Controller) Finalize(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	stored := machine.DeepCopy()
+	if !controllerutil.ContainsFinalizer(machine, v1alpha5.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	if err := c.cleanupNodeForMachine(ctx, machine); err != nil {
+		if termination.IsNodeDrainError(err) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("provider-id", machine.Status.ProviderID))
+	if err := c.cloudProvider.Delete(ctx, machine); cloudprovider.IgnoreMachineNotFoundError(err) != nil {
+		return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+	}
+	controllerutil.RemoveFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(stored, machine) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("removing machine termination finalizer, %w", err))
+		}
+		logging.FromContext(ctx).Infof("deleted machine")
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(ctx context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1alpha5.Machine{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &v1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+				node := o.(*v1.Node)
+				machineList := &v1alpha5.MachineList{}
+				if err := c.kubeClient.List(ctx, machineList, client.MatchingFields{"status.providerID": node.Spec.ProviderID}, client.Limit(1)); err != nil {
+					return []reconcile.Request{}
+				}
+				return lo.Map(machineList.Items, func(m v1alpha5.Machine, _ int) reconcile.Request {
+					return reconcile.Request{
+						NamespacedName: client.ObjectKeyFromObject(&m),
+					}
+				})
+			}),
+		).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 50})) // higher concurrency limit since we want fast reaction to node syncing and launch
+}
+
+func (c *Controller) cleanupNodeForMachine(ctx context.Context, machine *v1alpha5.Machine) error {
+	node, err := nodeForMachine(ctx, c.kubeClient, machine)
+	if err != nil {
+		// We don't clean the node if we either don't find a node or have violated the single machine to single node invariant
+		if IsNodeNotFoundError(err) || IsDuplicateNodeError(err) {
+			return nil
+		}
+		return err
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("node", node.Name))
+	if err = c.terminator.Cordon(ctx, node); err != nil {
```

Comment:
I'm not sure we should be doing this directly from within the machine controller.  Can we decouple this with a label so that all terminations go through the termination controller?

Reply:
We are moving the "termination controller" logic to the machine eventually, with a backdoor to be able to do a delete on the Node, which will trigger the termination finalization flow on the Machine, which will own the cordon, drain, delete flow right now. Can't remove it from the termination controller yet since this PR isn't actually doing the migration yet.